[Security] Exclude remember_me from default login authenticators

santysisi opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #60221 
| License       | MIT


🐛 Bugfix Description:
This fix addresses an issue where the remember_me authenticator was incorrectly included in the list of authenticators used during the security login process. When only one authenticator was configured and it was remember_me, the login flow could break or behave unexpectedly, as remember_me is not intended to be used as a primary authenticator. The login() method has been updated to exclude remember_me from the list, ensuring proper handling of login attempts and avoiding unintended authentication behavior.

santysisi commented:
I think (though I could be wrong) that the test failures are due to flaky tests.

94noni commented on src/Symfony/Bundle/SecurityBundle/Resources/config/security.php:
```diff
@@ -97,6 +97,7 @@
                     'security.csrf.token_manager' => service('security.csrf.token_manager')->ignoreOnInvalid(),
                 ]),
                 abstract_arg('authenticators'),
+                abstract_arg('defaultAuthenticators'),
```
Without leading « s » no?
As per the config 

Edit: wrong read lol

chalasr commented:
Thanks for the PR. As said in the linked issue, I'm rather 👎  on this as the proposed `default_authenticator` option doesn't have any effect outside the context of a `Security::login()` call and its name doesn't convey this which makes it confusing to me. Even if we make it obvious in the option name, I feel like a global option is not appropriate here as `login()` the use of marginal in an app.

santysisi commented:
Thanks for your feedback @chalasr , and you're right, I should have consulted you before proceeding with the implementation. I understand your concerns about approach

Do you think there's a different approach that might still address the underlying need, or would it make more sense to close this pull request?

chalasr commented:
If you agree to work on the suggested alternative which is to patch the `login()` method to basically make it exclude `remember_me` from the retrieved list of authenticators so it works seamlessly in case there's only one authenticator with rememberme configured, I'm fine with repurposing this very PR.

santysisi commented:
Yeah, I'm fine with that alternative, so I'll go ahead and make the change next weekend.
Thank you for your suggestion.

santysisi committed:
Fix: exclude remember_me from security login authenticators

santysisi commented:
Hi @chalasr! 👋

I've updated the branch with the patch to the login() method. Since this is a patch to the login behavior, I changed the base branch from 7.3 to 6.4, and also updated the title and description of the PR to reflect the changes.

If there's anything else I need to address, please let me know!

Thanks ❤️

rosier commented:
Would it be an option to always pick the first authenticator from the retrieved list of authenticators and not just when there is only one authenticator configured?  (Basically remove these lines https://github.com/symfony/symfony/blob/7.3/src/Symfony/Bundle/SecurityBundle/Security.php#L179-L181)

chalasr commented on src/Symfony/Bundle/SecurityBundle/Security.php:
```diff
@@ -70,6 +70,9 @@ class Security extends InternalSecurity implements AuthorizationCheckerInterface
     public const LAST_USERNAME = SecurityRequestAttributes::LAST_USERNAME;
 
+    /** @var string[] */
+    private const SECURITY_LOGIN_EXCLUSIONS = ['remember_me'];
```
let's remove this constant in favor of an inline `'remember_me' ===` check and reconsider if we have more than just remember_me someday

santysisi replied:
Hi @chalasr, thanks for your comment :smile:  I’ve made the change and removed the constant in favor of the inline `'remember_me' ===` check.

chalasr commented:
> I was also thinking about adding a new condition [here](https://github.com/symfony/symfony/blob/7.3/src/Symfony/Bundle/SecurityBundle/Security.php#L110) to check if the passed authenticatorName is remember_me. If so, it could trigger a deprecation notice in 7.4, and throw an error in version 8.0. What do you think about that approach?

I get your point but I don't think it is worth it right now, we can rethink about it the day it happens

chalasr reviewed (approved):
Authenticating through the `remember_me` authenticator was never meant to work at it's a very special one. 👍 to remove the ambiguity as a bugfix

fabpot commented:
Thank you @santysisi.
